### PR TITLE
Victor/7932 UP accepts bulk FHIR bundles

### LINKED ISSUE #7932: UP accepts submissions of FHIR bundles in bulk
### User Story:
As sender, I want to be able to send ReportStream multiple FHIR bundles in the same submission, so I can bulk upload reports.

### Description/Use Case
Simple Reports wants to be able to upload multiple reports in the same submission via our API. We need to implement receiving FHIR bundles as NDJSON to receive multiple bundles at the same time.  See the design document on #7401.

### Risks/Impacts/Considerations
* What is the impact in performance to the reports API when submitting many bundles?  Can we finish by 30 secs, which is the limit on Azure?  We are pretty sure this is a YES
* What happens when the payload is bigger than the allotted 100MB set by Azure? I think we just get some 400 or 500 error, but test this to inform senders what to expect. 

### Dev Notes:
* We will be using NDJSON to receive the bundles.
* You must validate each bundle (right now this is just for structure) and reject the payload if any are bad.
* You must be able to handle just one FHIR bundle as before, so use the content type as a trigger to which format you are receiving.
* The Jackson library seems to have support for NDJSON, but unsure if this makes sense over a simple split by newline.

### Acceptance Criteria
* ReportStream is capable of receiving a submission with an NDJSON payload that contains one or more FHIR bundles.
* ReportStream returns a report error if any FHIR bundle in the NDJSON payload cannot be read as a FHIR bundle (check for structure/syntax).
* ReportStream returns a good report response if all the FHIR bundles in the NDJSON can be read as FHIR bundles (check for structure/syntax).
* ReportStream debatches the submission into separate FHIR bundles during the convert step of the UP pipeline.

### PR DESCRIPTION
**Pull request by victor-chaparro**
This PR allows the Universal Pipeline to accept and ingest bulk FHIR bundles through the reports API endpoint. To be able to send bulk FHIR bundles the new content type application/fhir+ndjson had to be added. Ndjson stands for New line Delimited JSON and allows us to send multiple FHIR bundles new line delimited in one message. Previous smoke tests using FHIR data had to be modified to remove any new lines that would break the tests.

Test Steps:

1. Send a POST request to send a bulk FHIR report. Sample request: 
[Tests.postman_collection.txt](https://github.com/CDCgov/prime-reportstream/files/10572145/Tests.postman_collection.txt)

2. Verify that the request was accepted

3. Call the submission history API to verify the reports were delivered: http://localhost:7071/api/waters/report/{{reportId}}/history
4. Optional: Check sftp folder to verify that the HL7 message get routed correctly

## Changes
- Updated the FHIR format in both Sender.kt and Report.kt to support bulk FHIR bundles (application/fhir+ndjson)
- Added a new FHIRReader.kt class that takes in a raw fhir+ndjson message and returns a list of bundles
- Updated ELRReceiver.kt to be able to support sending and receving bulk FHIR bundles
- Fixed a bug on FHIRConverter.kt that didn't allow for multiple bundles to be sent to Routing
- Fixed a bug on SubmissionHistory.kt that would create the same destination multiple times in the API response
- Updated unit tests to use new fhir+ndjson message format

### Testing
- [x] Tested locally?
- [x] Ran `./prime test` or `./gradlew testSmoke` against local Docker ReportStream container?
- [x] Added tests?

## Linked Issues
- Fixes #7932

### CODE REVIEW

**Comment by victor-chaparro** [comment]
Testing notification

File: prime-router/src/test/kotlin/SubmissionReceiverTests.kt [hunk 1]
```diff
@@ -301,6 +301,11 @@ class SubmissionReceiverTests {
         "42&ISO||445297001^Swab of internal nose^SCT^^^^2.67||||53342003^Internal nose structure (body structure)^" +
         "SCT^^^^2020-09-01|||||||||202108020000-0500|20210802000006.0000-0500"
 
+    val bulk_fhir_bundles = """
```
Comment (PatriciaPerozoUSDS) [inline prime-router/src/test/kotlin/SubmissionReceiverTests.kt:304]: In FhirConverterTests.kt we're reading FHIR bundles in from external files. I think it makes the test easier to read. Would it be possible to switch the rest of the tests over to using that strategy as well?

File: prime-router/src/main/kotlin/history/SubmissionHistory.kt [hunk 2]
```diff
@@ -336,9 +336,19 @@ class DetailedSubmissionHistory(
         }
 
         // Grab destinations from the "translate" action, if the submission made it that far
-        descendants.filter { it.actionName == TaskAction.translate }.forEach { descendant ->
+        descendants.filter { it.actionName == TaskAction.translate }.distinct().forEach { descendant ->
             descendant.destinations.forEach { dest ->
```
Comment (PatriciaPerozoUSDS) [inline prime-router/src/main/kotlin/history/SubmissionHistory.kt:340]: It might be useful to distinguish between the Submission History destinations property and dest the destination within the list of descendant destinations. I was trying to come up with variable names and honestly the only one that seemed descriptive was descendantDest or descendantDestination. 

Feel free to leave it as is if others aren't confused by it

File: prime-router/src/main/kotlin/fhirengine/utils/FHIRReader.kt [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+import java.io.StringReader
+
+/**
+ * Converts raw FHIR data (ndjson Newline Delimited Json) to Bundle objects.
+ */
+class FHIRReader(private val actionLogger: ActionLogger) : Logging {
+    /**
+     * Returns one or more bundles read from the raw FHIR data.
+     */
+    fun getBundles(rawMessage: String): List<Bundle> {
```
Comment (PatriciaPerozoUSDS) [inline prime-router/src/main/kotlin/fhirengine/utils/FHIRReader.kt:17]: Can you add the arguments to the kdoc description - Something like 
"Returns one or more bundles read from the FHIR data in [rawMessage]."

File: prime-router/src/main/kotlin/history/SubmissionHistory.kt [hunk 4]
```diff
@@ -336,9 +336,19 @@ class DetailedSubmissionHistory(
         }
 
         // Grab destinations from the "translate" action, if the submission made it that far
-        descendants.filter { it.actionName == TaskAction.translate }.forEach { descendant ->
+        descendants.filter { it.actionName == TaskAction.translate }.distinct().forEach { descendant ->
```
Comment (PatriciaPerozoUSDS) [inline prime-router/src/main/kotlin/history/SubmissionHistory.kt:339]: Talked about over slack - please remove the distinct here

**Review by PatriciaPerozoUSDS — APPROVED** [review]
This looks good to me aside from the nits. Thank you for pulling out the FHIRReader into a new class.

**Review by JessicaWNava — APPROVED** [review]
Looks good to me, just one question

File: prime-router/src/main/kotlin/Sender.kt [hunk 5]
```diff
@@ -124,7 +124,7 @@ abstract class Sender(
     enum class Format(val mimeType: String) {
         CSV("text/csv"),
         HL7("application/hl7-v2"),
-        FHIR("application/fhir+json"),
+        FHIR("application/fhir+ndjson"),
```
[the remark this answers is not in the document]
Comment (carlosfelix2) [reply prime-router/src/main/kotlin/Sender.kt:127]: It is a different format: JSON is one, NDJSON is another

File: prime-router/src/main/kotlin/fhirengine/utils/FHIRReader.kt [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+import gov.cdc.prime.router.InvalidReportMessage
+import org.apache.logging.log4j.kotlin.Logging
+import org.hl7.fhir.r4.model.Bundle
+import java.io.BufferedReader
+import java.io.StringReader
+
+/**
+ * Converts raw FHIR data (ndjson Newline Delimited Json) to Bundle objects.
+ */
+class FHIRReader(private val actionLogger: ActionLogger) : Logging {
```
Comment (carlosfelix2) [inline prime-router/src/main/kotlin/fhirengine/utils/FHIRReader.kt:13]: Merge this code with the FHIRTranscoder class. I think the name FHIRReader is fine, but the transcoder class kind of does some of the reading too.

File: prime-router/src/main/kotlin/fhirengine/utils/FHIRReader.kt (hunk 3 above)
Comment (carlosfelix2) [reply prime-router/src/main/kotlin/fhirengine/utils/FHIRReader.kt:17]: Yes, needs full kdoc. Missing @ return

File: prime-router/src/main/kotlin/fhirengine/utils/FHIRReader.kt [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+
+import gov.cdc.prime.router.ActionLogger
+import gov.cdc.prime.router.InvalidReportMessage
+import org.apache.logging.log4j.kotlin.Logging
+import org.hl7.fhir.r4.model.Bundle
+import java.io.BufferedReader
+import java.io.StringReader
+
+/**
+ * Converts raw FHIR data (ndjson Newline Delimited Json) to Bundle objects.
```
Comment (carlosfelix2) [inline prime-router/src/main/kotlin/fhirengine/utils/FHIRReader.kt:11]: Kdoc. Missing properties

File: prime-router/src/main/kotlin/fhirengine/engine/FHIRConverter.kt [hunk 8]
```diff
@@ -178,9 +180,10 @@ class FHIRConverter(
      * @return a list containing a FHIR bundle
      */
     internal fun getContentFromFHIR(
-        message: RawSubmission
+        message: RawSubmission,
+        actionLogger: ActionLogger
```
Comment (carlosfelix2) [inline prime-router/src/main/kotlin/fhirengine/engine/FHIRConverter.kt:184]: kdoc. Missing arg

File: prime-router/src/test/kotlin/SubmissionReceiverTests.kt (hunk 1 above)
Comment (victor-chaparro) [reply prime-router/src/test/kotlin/SubmissionReceiverTests.kt]: Moved test data to external files.

**Review by carlosfelix2 — CHANGES_REQUESTED** [review]

File: prime-router/src/main/kotlin/fhirengine/utils/FHIRReader.kt (hunk 6 above)
Comment (victor-chaparro) [reply prime-router/src/main/kotlin/fhirengine/utils/FHIRReader.kt]: Merged FHIRReader into FhirTranscoder

File: prime-router/src/test/kotlin/fhirengine/utils/FhirTranscoderTests.kt [hunk 9]
```diff
@@ -71,4 +75,47 @@ class FhirTranscoderTests {
         assertThat(decodedBundle).isNotNull()
         assertThat(decodedBundle).equals(testBundle)
```
Comment (carlosfelix2) [inline prime-router/src/test/kotlin/fhirengine/utils/FhirTranscoderTests.kt:76]: Wow! Can you please open a new issue to fix this. There are 5 instances of .equals use with assertTHat and it needs to be isEqualTo. Otherwise these tests do not test anything.

File: prime-router/src/main/kotlin/fhirengine/utils/FhirTranscoder.kt [hunk 10]
```diff
@@ -50,4 +55,39 @@ object FhirTranscoder : Logging {
...
+        } else {
+            val bufferedReader = BufferedReader(StringReader(rawMessage))
+            val iterator = bufferedReader.lineSequence().iterator()
+            var index = 1
+            while (iterator.hasNext()) {
+                try {
+                    val line = iterator.next()
+                    val bundle = decode(line)
+                    if (bundle.isEmpty) {
+                        actionLogger.error(InvalidReportMessage("$index: Unable to find FHIR Bundle in provided data."))
```
Comment (carlosfelix2) [inline prime-router/src/main/kotlin/fhirengine/utils/FhirTranscoder.kt:78]: Missing unit test if bundle is empty.

**Review by carlosfelix2 — APPROVED** [review]
